### PR TITLE
Add exact Qwen4Exp text core

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

- add the pinned `qwen4_exp_text` decoder for `Qwen/Qwen3.8-Flash-Next@f5d08274bafd880402bd16f5e3e6c514136ec06c` and Transformers `598d8ba8baaec7fec5a22da0e2844c7bf4ea20e1`
- implement the 3:1 Gated-DeltaNet/QSA schedule, hyper-connections, routed plus shared MoE, PLE hashing/convolution, and heterogeneous cache ABI
- preserve the BF16 checkpoint contract by keeping DeltaNet recurrence and recurrent state in float32
- make QSA exact for non-contiguous masks and validate official packed experts, deterministic hash buffers, and all PLE shards fail-closed
- register the standalone text route while rejecting unsupported multimodal export and dedicated MTP embeddings

## Validation

- targeted Qwen4Exp/DeltaNet/build/weight tests: `37 passed, 7 skipped, 1857 deselected`
- exact pinned Transformers reduced-model parity: prefill, token-by-token decode, and non-contiguous-mask parity passed
- broad serial suite: `8377 passed, 64 skipped, 12 deselected, 1 subtest passed`
- `lintrunner init && lintrunner f --output oneline --all-files && lintrunner -a`: no lint issues
- GPT-5.6 Sol medium review: no significant findings
- pinned checkpoint index verified: 31 ignored `mtp.*` tensors, 128 PLE shards, and official packed expert/shared-gate names

## Fail-closed guards

- rejects invalid heterogeneous schedules, linear head ratios/dimensions, QSA dimensions/budgets, PLE dimensions/sharding, non-float32 recurrent state, and dedicated MTP embeddings
- rejects malformed packed expert tensors, changed deterministic hash buffers, missing/extra PLE shards, and wrong reconstructed PLE table shapes
- composite `qwen4_exp` export requires `text_only=True`; multimodal export is not silently aliased

## Waivers

- Full-checkpoint L2/L4/L5 and CLI export were not run locally: the pinned model includes a roughly 95 GiB PLE embedding table and is not a practical local golden fixture. The independently instantiated pinned Transformers reduced model provides full-logit L3 parity.
- BF16 execution was structurally validated through the graph/state ABI. Stock local ORT CPU cannot execute BF16 MatMul and expands standard BF16 Attention through an unsupported mixed-type internal Add; no CUDA runner was available.
- The MTP sidecar and multimodal wrapper are intentionally out of scope. The pinned official causal runtime ignores `mtp.*`; unsupported contracts fail closed rather than exporting approximate behavior.

### CODE REVIEW

File: src/mobius/_configs/_base.py [hunk 1]
```diff
@@ -1531,6 +1531,229 @@ class CausalLMConfig(ArchitectureConfig):
...
+            raise ValueError(f"Qwen4-Exp requires hc_count > 1, got {self.hc_count}")
+        if self.hc_lowrank <= 0:
+            raise ValueError(f"Qwen4-Exp requires hc_lowrank > 0, got {self.hc_lowrank}")
+        if self.mamba_ssm_dtype != ir.DataType.FLOAT:
+            raise ValueError(
+                "Qwen4-Exp requires mamba_ssm_dtype=float32 for the pinned "
+                "Gated-DeltaNet recurrence"
+            )
+        if not self.num_local_experts or self.num_local_experts <= 0:
+            raise ValueError("Qwen4-Exp num_experts must be > 0")
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/_configs/_base.py]: Fixed in 5b04538f: both validation messages now consistently name the normalized `num_local_experts` field.

File: src/mobius/models/qwen4_exp.py [hunk 2]
```diff
@@ -0,0 +1,1158 @@
...
+
+    def forward(
+        self,
+        op: OpBuilder,
+        input_ids: ir.Value,
+        attention_mask: ir.Value,
+        position_ids: ir.Value,
+        past_position_ids: ir.Value,
+        past_key_values: list[tuple[ir.Value, ...]],
+    ):
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/qwen4_exp.py:1098]: Resolved in 5b04538f without inventing a fallback position history. Qwen4-Exp's task requires `past_position_ids` for exact QSA cache semantics, so the model now subclasses `nn.Module` rather than claiming substitutability with `CausalLMModel.forward`. Making the input optional would fail open for cached/custom-position calls.

File: src/mobius/models/qwen4_exp.py [hunk 3]
```diff
@@ -0,0 +1,1158 @@
...
+from mobius.components import (
+    Embedding,
+    Linear,
+    SoftmaxTopKGate,
+    create_attention_bias,
+    get_activation,
+    initialize_rope,
+)
+from mobius.components._attention import Qwen35Attention
+from mobius.components._gated_deltanet import GatedDeltaNet
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/qwen4_exp.py]: Fixed in 5b04538f: `Qwen35Attention` and `GatedDeltaNet` now come from the public `mobius.components` API.

File: src/mobius/integrations/transformers/_qwen4_exp_weights.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
+# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/transformers/_qwen4_exp_weights.py:1]: Fixed in 17084733 by applying the pinned Ruff formatter; full local lintrunner now passes.

File: src/mobius/integrations/transformers/_qwen4_exp_weights_test.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+import onnx_ir as ir
+import safetensors.torch
+import torch
+
+from mobius._builder import build_from_module
+from mobius.models.qwen4_exp import Qwen4ExpCausalLMModel
+from mobius.models.qwen4_exp_test import _config
+
+from ._qwen4_exp_weights import stream_qwen4_exp_safetensors_to_model
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/transformers/_qwen4_exp_weights_test.py]: Fixed in 17084733: the test now imports the Qwen4 streaming loader through its absolute package path.

File: src/mobius/integrations/transformers/_qwen4_exp_weights_test.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
+# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/transformers/_qwen4_exp_weights_test.py:1]: Fixed in 17084733 by applying the pinned Ruff formatter to the streaming-loader test.
